### ISSUE TITLE
Wrongly generated .d.ts

### ISSUE THREAD

**Issue by jerrygreen** [AssemblyScript/assemblyscript#1704]
I've found a good issue [here](https://github.com/AssemblyScript/assemblyscript/issues/1488), and figured out that I can actually generate `.d.ts` by setting `tsdFile` option in config, and use AssemblyScript types inside TypeScript! Which is absolutely cool, but I found that automatically it actually generates slightly wrong hierarchy.

I have further AssemblyScript code:

```ts
export function add(a: i32, b: i32): i32 {
  return a + b
}
```

And a wasm module as a react component:

```tsx
const AsmComponent = dynamic({
  loader: async () => {
    const wasmModule = await import('../../../asm/build/optimized') // <- imports `.wasm`

    return (props) => <>{wasmModule.add(1, 2)}</>
  },
  ssr: false,
})
```

And I render it in React like so:

```tsx
          <p>Execute function from wasm module:</p>
          <p>
            1 + 2 = <AsmComponent />
          </p>
```

And it works:

![image](https://user-images.githubusercontent.com/13215662/109417458-950c1200-79e5-11eb-8e12-6816ccdd48fb.png)

So it works, but look at my types:

> Property 'add' does not exist on type 'typeof import("c:/Users/Jerry/projects/project-x/asm/build/optimized")'

![image](https://user-images.githubusercontent.com/13215662/109417510-c5ec4700-79e5-11eb-8939-d225f6879546.png)

Now I can fix these types by replacing `asmModule.add` into `asmModule.default.add`:

![image](https://user-images.githubusercontent.com/13215662/109417560-fe8c2080-79e5-11eb-98d8-5fb797444a91.png)

But if I now run actual application, "default" doesn't actually have "add" function:

![image](https://user-images.githubusercontent.com/13215662/109417609-31ceaf80-79e6-11eb-9a88-01dfeb9c3ca9.png)

My automatically generated `optimized.d.ts` looks like so:

```ts
declare module ASModule {
  type i8 = number;
  type i16 = number;
  type i32 = number;
  type i64 = bigint;
  type isize = number;
  type u8 = number;
  type u16 = number;
  type u32 = number;
  type u64 = bigint;
  type usize = number;
  type f32 = number;
  type f64 = number;
  type bool = boolean | number;
  export function add(a: i32, b: i32): i32;
}
export default ASModule;
```

If I manually remove the namespace, leaving bare types like so:

```ts
type i8 = number;
type i16 = number;
type i32 = number;
type i64 = bigint;
type isize = number;
type u8 = number;
type u16 = number;
type u32 = number;
type u64 = bigint;
type usize = number;
type f32 = number;
type f64 = number;
type bool = boolean | number;
export function add(a: i32, b: i32): i32;
```

Then both types and actual code go aligned, and both working!

Pls can someone change how `.d.ts` file is generated, and generate it without a namespace?

**Comment by dcodeIO** [comment]
The bundler seems to make an assumption about how a `.wasm` file maps to a `.d.ts`, yet it may well be that AssemblyScript is the only tool currently emitting a `.d.ts` for a `.wasm` file (not 100% sure though), and it recommends to use it [with the loader](https://www.assemblyscript.org/loader.html#typescript-definitions). I am not opposed to change the way this works, however, especially with [esm-integration](https://github.com/WebAssembly/esm-integration/tree/master/proposals/esm-integration) looming, but since this is a breaking change it would be good to evaluate if there are any clear signals for this case from standards bodies, or TS for that matter, yet, so it becomes unlikely that we'll have to change this again in the future.

**Comment by JerryGreen** [comment]
@dcodeIO this loader seems to be some glue code for environments where dynamic imports of wasm modules doesn't work and require fetching the module first, maybe some older node versions, maybe for esm modules, or idk... But you can clearly see that no any glue code required here (if anyone interested, I used Node 14.15.2 and Typescript 4.1.5), so this loader is kinda unneeded (again, I admit it still might be needed for some other environments though).

Anyway, from what I see from VSCode right now, the generated types are not aligned with actual values here. If I decided to fix how imports work, to change the value structure to make them fit for existing types, - this would be a hugely breaking change, since that would change how code works. But this particular fix does not change how code works, and only changes how types are being generated, so even though breaking, it's very little breaking change, and barely someone will notice that. For starters, it is an optional feature that is disabled by default and has to be enabled in `asconfig.json`:

```jsonc
{
  "targets": {
    "release": {
      "tsdFile": "src/build/optimized.d.ts" // <--- not generated if not set
    }
  }
}
```

Another point: it's not even featured in [docs](https://www.assemblyscript.org/introduction.html) (I myself found it in [this issue](https://github.com/AssemblyScript/assemblyscript/issues/1488), and otherwise wouldn't know it's possible to generate `.ts.d` file):

![image](https://user-images.githubusercontent.com/13215662/109498869-eb488600-7ab5-11eb-99a4-2676ef7c3112.png)

I agree that people that are using `tsdFile` feature, - their AsmModule types will broke upon updating AssemblyScript compiler. But `asc`'s major version is still currently 0, so I assume people might expect something like that, and be ready to fix that upon updating. For the rest people, like who are starting new AssemblyScript projects, - for them it will simplify the developer experience because no loader required. So to me it looks like: "omg let's get this merged asap" actually.

I'm not sure what you expect from me particularly... To ping some people from TS team or something? I don't get...

**Comment by JerryGreen** [comment]
@dcodeIO I just found those imports don't work in pure Node with default flags though, - node doesn't support import syntax for wasm modules. But if you pass `--experimental-wasm-modules` argument to node, then it will work. Though in my particular environment I use NextJS and I suppose it has webpack config to make that happen rather than node experimental flag, but they both seem to use the same approach. In this article you can find code samples with the same exact code and structure:

https://www.joyent.com/blog/improved-wasm-support-coming-to-node#importing-webassembly-modules

```
'use strict';
(async () => {
  const example = await import('./example.wasm');

  console.log(example.addTwo(8, 5));
})();
```

So look, it worked like years ago (like a couple years ago, maybe more?). So I guess this structure is right. Idk why current declaration file wraps it into some kind of namespace (confusingly naming it "module" too), it doesn't make sense. No wrappers needed

P.S. Sorry for some kind of pressure, I just really want it to land, so I can easily `yarn add assemblyscript`, and have comfort tinkering things while having auto-generated types that I don't have to fix manually

**Comment by JerryGreen** [comment]
@dcodeIO I see... I [added](https://github.com/AssemblyScript/assemblyscript/commit/26021cfb980d16c8472e754b1626cc586a11f415) to the [existing PR](https://github.com/AssemblyScript/assemblyscript/pull/1705) a couple integrated tests, that do the whole cycle:

1. Use existing `asc` to compile simple `add` module into `.wasm`
2. Use `-d` (`--tsdFile`) option to generate according `.wasm.d.ts` (using the new structure of `.d.ts` file)
3. Have typescript file that is using this `.wasm` module
4. Have typescript compiler involved to assure it can find corresponding types (it will fail if it won't find it)
5. Execute generated javascript to make sure the `.wasm` module works
6. And all that is included into `npm run test` command, to assure those imports will work in the future, both js functionality and ts declarations

One thing to note that `.d.ts` file should have exactly the same name as `.wasm` module, with added `.d.ts` in the end, so VSCode is able to find those declarations.

**Comment by JerryGreen** [comment]
@dcodeIO any info about when it can be merged? It's been two days, and you didn't answer anything... Only quietly added this "breaking change" label 😂

**Comment by dcodeIO** [comment]
I am still rather in favor of making your proposed change, so far marking it to potentially include in the next breaking window for a 0.19, ideally along other similar breaking changes.

**Comment by dcodeIO** [comment]
Random idea: Could we make this change so it would be backwards compatible, via the `default` export re-exporting everything in the file so it will still work like before?

**Comment by JerryGreen** [comment]
@dcodeIO I don’t think it’s sensible to bear this wrong yet backwards compatible code. It works just wrong way, not how it should work.

I think it is possible to make some improvements though. Let me explain... Currently whether you export function via `default` or not, the module will be working the same way anyway, i.e. generated wasm won’t change, it will always be imported like so:

```ts
const { add } = await import('/modules/my-module.wasm')
add // <— exported function
```

It is possible to distinguish default imports vs non-default imports and for default imports make it work this way:

```ts
const { default: add, multiply, etc } = await import('/modules/my-module.wasm')
add // <— exported function via default keyword
```

Similarly to [how it’s working in js right now](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import):

```ts
const { default: myDefault, foo, bar } = await import('/modules/my-module.js');
myDefault // <— also exported via default keyword
```

I think it would be ideal... To make AssemblyScript more aligned to how JavaScript work (where possible). But I guess it will involve some work for more compiler logic rather than just generating `.d.ts` file, to how wasm generated itself. So:

1. I think it will be even more breaking change rather than just changing how `.d.ts` is generated. Because it will break not just developer experience regarding types, but how code itself works (since it will generate different structure in some cases)
2. I’m not ready to do such changes. I’m not even sure where to do these changes honestly.

Btw, types are generated based on if there’s default export or not ALREADY, i.e. there will be slightly different types in `.d.ts`, - so both approaches are already supported by generated `.d.ts`. But again I’m not ready to make changes to how actual compiler works regarding those imports. Only types right now. If anyone would like to change how default imports actually work, and make it more aligned to JavaScript, - he can do another PR.

**Comment by JerryGreen** [comment]
I actually figured out that I can just use:

```json
"assemblyscript": "https://github.com/jerrygreen/assemblyscript/tarball/issue1704",
```

In my `package.json`, so compiler can generate proper `.d.ts`. Sorry for any pressure before: take any time to merge this whenever you think you should do that 😏

**Comment by dcodeIO** [comment]
Fixed in #1872 :)
